File: python/advanced.py
"""
bench_advanced.py — Advanced Python Benchmark
Tests OOP dispatch, recursion, memory/GC pressure, generator pipelines,
comprehensions, and lambda/closure overhead.
Run native vs. obfuscated (PyArmor, etc.) to isolate instrumentation costs.
"""

import time
import math
import sys

def clock():
    return time.perf_counter()

script_start = clock()
sys.setrecursionlimit(50_000)

print("=== Advanced Benchmark (Python) ===")
print("Testing: OOP, recursion, memory churn, generators, comprehensions, closures")
print()

# ── Test 1: OOP Method Dispatch ───────────────────────────────────────────────
print("Test 1: OOP Method Dispatch (500K calls)")

class Animal:
    __slots__ = ("name", "hp", "atk", "xp")

    def __init__(self, name, hp, atk):
        self.name = name
        self.hp   = hp
        self.atk  = atk
        self.xp   = 0

    def attack(self, other):
        dmg = max(1, self.atk - int(other.hp * 0.05))
        other.hp -= dmg
        self.xp  += dmg
        return dmg

t1 = clock()
wolf  = Animal("Wolf",  999_999, 10)
sheep = Animal("Sheep", 999_999,  1)
dmg_sink = 0
for _ in range(500_000):
    dmg_sink += wolf.attack(sheep)
t1 = clock() - t1
print(f"  Damage: {dmg_sink}")
print(f"  Time:   {t1:.4f}s")
print()

# ── Test 2: Inheritance Chain (3 levels) ──────────────────────────────────────
print("Test 2: 3-Level Inheritance Chain (300K calls)")

class Base:
    def __init__(self, v):    self.value = v
    def get(self):            return self.value

class Mid(Base):
    def __init__(self, v, m): super().__init__(v); self.mult = m
    def scaled(self):         return self.get() * self.mult

class Top(Mid):
    def __init__(self, v, m, s): super().__init__(v, m); self.shift = s
    def final(self):             return self.scaled() + self.shift

t2 = clock()
obj = Top(3, 7, 2)
chain_sink = 0
for _ in range(300_000):
    chain_sink += obj.final()
t2 = clock() - t2
print(f"  Sink: {chain_sink}")
print(f"  Time: {t2:.4f}s")
print()

# ── Test 3: Recursive Fibonacci (naive) ───────────────────────────────────────
print("Test 3: Naive Fibonacci fib(30), 300 times")

def fib(n):
    if n <= 1: return n
    return fib(n - 1) + fib(n - 2)

t3 = clock()
fib_sink = sum(fib(30) for _ in range(300))
t3 = clock() - t3
print(f"  fib(30) = {fib(30)}")
print(f"  Sink:    {fib_sink}")
print(f"  Time:    {t3:.4f}s")
print()

# ── Test 4: Memory Churn (alloc + GC) ─────────────────────────────────────────
print("Test 4: Dict Churn (200K alloc+abandon cycles)")
t4 = clock()
churn_sink = 0
for i in range(1, 200_001):
    d = {"a": i, "b": i * 2, "c": i * 3, "d": i * 4}
    churn_sink += d["a"] + d["d"]
t4 = clock() - t4
print(f"  Sink: {churn_sink}")
print(f"  Time: {t4:.4f}s")
print()

# ── Test 5: List Comprehensions vs Loops ──────────────────────────────────────
print("Test 5: List Comprehensions (1K lists of 1K elements each)")
t5 = clock()
comp_sink = 0
for i in range(1, 1_001):
    lst = [j * i + (j % 7) for j in range(1_000)]
    comp_sink += lst[0] + lst[-1]
t5 = clock() - t5
print(f"  Sink: {comp_sink}")
print(f"  Time: {t5:.4f}s")
print()

# ── Test 6: Generator Pipeline ────────────────────────────────────────────────
print("Test 6: Generator Pipeline (500K items through 3-stage pipeline)")

def source(n):
    for i in range(1, n + 1):
        yield i

def stage_filter(it):
    for x in it:
        if x % 3 != 0:
            yield x

def stage_transform(it):
    for x in it:
        yield x * x - x + 1

t6 = clock()
pipeline = stage_transform(stage_filter(source(500_000)))
gen_sink = sum(pipeline)
t6 = clock() - t6
print(f"  Sink: {gen_sink}")
print(f"  Time: {t6:.4f}s")
print()

# ── Test 7: Closure / Lambda Overhead ─────────────────────────────────────────
print("Test 7: Closure Factory (300K closures created + called)")

def make_adder(n):
    return lambda x: x + n

t7 = clock()
closure_sink = 0
for i in range(1, 300_001):
    adder = make_adder(i)
    closure_sink += adder(1)
t7 = clock() - t7
print(f"  Sink: {closure_sink}")
print(f"  Time: {t7:.4f}s")
print()

# ── Test 8: Merge Sort (2K arrays × 200 elements) ─────────────────────────────
print("Test 8: Merge Sort (2K arrays of 200 elements)")

def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left  = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    result, i, j = [], 0, 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i]); i += 1
        else:
            result.append(right[j]); j += 1
    result.extend(left[i:]); result.extend(right[j:])
    return result

def lcg(seed):
    s = seed
    while True:
        s = (s * 1_664_525 + 1_013_904_223) & 0xFFFFFFFF
        yield s % 10_000

t8 = clock()
sort_sink = 0
for i in range(1, 2_001):
    gen = lcg(i * 31_337)
    arr = [next(gen) for _ in range(200)]
    sorted_arr = merge_sort(arr)
    sort_sink += sorted_arr[0] + sorted_arr[-1]
t8 = clock() - t8
print(f"  Sink: {sort_sink}")
print(f"  Time: {t8:.4f}s")
print()

# ── Summary ───────────────────────────────────────────────────────────────────
script_total = clock() - script_start

print("=== Summary ===")
print(f"OOP Dispatch:   {t1:.4f}s")
print(f"Inherit Chain:  {t2:.4f}s")
print(f"Fibonacci:      {t3:.4f}s")
print(f"Dict Churn:     {t4:.4f}s")
print(f"Comprehensions: {t5:.4f}s")
print(f"Generator Pipe: {t6:.4f}s")
print(f"Closures:       {t7:.4f}s")
print(f"Merge Sort:     {t8:.4f}s")
print("---------------")
print(f"TOTAL SCRIPT:   {script_total:.4f}s")
print()
print(f"RESULTS_CSV:{t1},{t2},{t3},{t4},{t5},{t6},{t7},{t8},{script_total}")

File: javascript/advanced.js
/**
 * bench_advanced.js — Advanced JavaScript Benchmark
 * Run with: node bench_advanced.js
 * Tests: prototype/class dispatch, closure factories, recursion,
 *        Map/Set vs plain objects, array methods, and Promise overhead.
 * Compare native vs obfuscated builds to isolate instrumentation costs.
 */

"use strict";

function clock() {
  return Number(process.hrtime.bigint()) / 1e9;
}

const scriptStart = clock();

console.log("=== Advanced Benchmark (JavaScript) ===");
console.log("Testing: class dispatch, prototype chains, closures, recursion, Map/Set, array methods, async");
console.log();

// ── Test 1: Class Method Dispatch (500K calls) ────────────────────────────────
console.log("Test 1: Class Method Dispatch (500K calls)");

class Animal {
  constructor(name, hp, atk) {
    this.name = name;
    this.hp   = hp;
    this.atk  = atk;
    this.xp   = 0;
  }
  attack(other) {
    const dmg = Math.max(1, this.atk - Math.floor(other.hp * 0.05));
    other.hp -= dmg;
    this.xp  += dmg;
    return dmg;
  }
}

let t1 = clock();
const wolf  = new Animal("Wolf",  999_999, 10);
const sheep = new Animal("Sheep", 999_999,  1);
let dmgSink = 0;
for (let i = 0; i < 500_000; i++) dmgSink += wolf.attack(sheep);
t1 = clock() - t1;
console.log(`  Damage: ${dmgSink}`);
console.log(`  Time:   ${t1.toFixed(4)}s`);
console.log();

// ── Test 2: Prototype Inheritance Chain (3 levels, 300K calls) ────────────────
console.log("Test 2: 3-Level Prototype Chain (300K calls)");

class Base {
  constructor(v)      { this.value = v; }
  get()               { return this.value; }
}
class Mid extends Base {
  constructor(v, m)   { super(v); this.mult = m; }
  scaled()            { return this.get() * this.mult; }
}
class Top extends Mid {
  constructor(v, m, s){ super(v, m); this.shift = s; }
  final()             { return this.scaled() + this.shift; }
}

let t2 = clock();
const obj = new Top(3, 7, 2);
let chainSink = 0;
for (let i = 0; i < 300_000; i++) chainSink += obj.final();
t2 = clock() - t2;
console.log(`  Sink: ${chainSink}`);
console.log(`  Time: ${t2.toFixed(4)}s`);
console.log();

// ── Test 3: Closure Factory (300K closures created + called) ─────────────────
console.log("Test 3: Closure Factory (300K closures)");

function makeAdder(n) { return x => x + n; }

let t3 = clock();
let closureSink = 0;
for (let i = 1; i <= 300_000; i++) {
  const adder = makeAdder(i);
  closureSink += adder(1);
}
t3 = clock() - t3;
console.log(`  Sink: ${closureSink}`);
console.log(`  Time: ${t3.toFixed(4)}s`);
console.log();

// ── Test 4: Recursive Fibonacci (naive, 300 times) ───────────────────────────
console.log("Test 4: Naive Fibonacci fib(30), 300 times");

function fib(n) { return n <= 1 ? n : fib(n - 1) + fib(n - 2); }

let t4 = clock();
let fibSink = 0;
for (let i = 0; i < 300; i++) fibSink += fib(30);
t4 = clock() - t4;
console.log(`  fib(30) = ${fib(30)}`);
console.log(`  Sink:    ${fibSink}`);
console.log(`  Time:    ${t4.toFixed(4)}s`);
console.log();

// ── Test 5: Map vs Object (100K inserts + reads each) ────────────────────────
console.log("Test 5: Map vs Plain Object (100K ops each)");

let t5 = clock();
// Map
const map = new Map();
for (let i = 1; i <= 100_000; i++) map.set(`key_${i}`, i * 3);
let mapSink = 0;
for (let i = 1; i <= 100_000; i++) mapSink += map.get(`key_${i}`);
// Plain object
const pobj = Object.create(null);
for (let i = 1; i <= 100_000; i++) pobj[`key_${i}`] = i * 3;
let pobjSink = 0;
for (let i = 1; i <= 100_000; i++) pobjSink += pobj[`key_${i}`];
t5 = clock() - t5;
console.log(`  Map sink:  ${mapSink}`);
console.log(`  Obj sink:  ${pobjSink}`);
console.log(`  Time:      ${t5.toFixed(4)}s`);
console.log();

// ── Test 6: Array Method Pipeline (map → filter → reduce, 1K × 1K arrays) ────
console.log("Test 6: Array Method Pipeline (1K arrays × 1K elements)");

let t6 = clock();
let pipeSink = 0;
for (let i = 1; i <= 1_000; i++) {
  const arr = Array.from({ length: 1_000 }, (_, j) => j * i + (j % 7));
  const result = arr
    .map(x => x * 2 + 1)
    .filter(x => x % 3 !== 0)
    .reduce((acc, x) => acc + x, 0);
  pipeSink += result;
}
t6 = clock() - t6;
console.log(`  Sink: ${pipeSink}`);
console.log(`  Time: ${t6.toFixed(4)}s`);
console.log();

// ── Test 7: Set Operations (union / intersection, 10K rounds) ─────────────────
console.log("Test 7: Set Union & Intersection (10K rounds)");

let t7 = clock();
let setSink = 0;
for (let i = 1; i <= 10_000; i++) {
  const a = new Set(Array.from({ length: 100 }, (_, j) => (j * i) % 200));
  const b = new Set(Array.from({ length: 100 }, (_, j) => (j * (i + 1)) % 200));
  // union size
  const union = new Set([...a, ...b]);
  // intersection size
  let inter = 0;
  for (const v of a) if (b.has(v)) inter++;
  setSink += union.size + inter;
}
t7 = clock() - t7;
console.log(`  Sink: ${setSink}`);
console.log(`  Time: ${t7.toFixed(4)}s`);
console.log();

// ── Test 8: Async/Promise Overhead (100K resolved promises) ──────────────────
console.log("Test 8: Promise Overhead (100K sequential awaits)");

async function runPromises() {
  let promiseSink = 0;
  const t8start = clock();
  for (let i = 1; i <= 100_000; i++) {
    const v = await Promise.resolve(i * 2);
    promiseSink += v;
  }
  const t8 = clock() - t8start;
  console.log(`  Sink: ${promiseSink}`);
  console.log(`  Time: ${t8.toFixed(4)}s`);
  console.log();
  return t8;
}

// Run everything async so Test 8 can use await
(async () => {
  const t8 = await runPromises();
  const scriptTotal = clock() - scriptStart;

  console.log("=== Summary ===");
  console.log(`Class Dispatch:   ${t1.toFixed(4)}s`);
  console.log(`Prototype Chain:  ${t2.toFixed(4)}s`);
  console.log(`Closure Factory:  ${t3.toFixed(4)}s`);
  console.log(`Fibonacci:        ${t4.toFixed(4)}s`);
  console.log(`Map vs Object:    ${t5.toFixed(4)}s`);
  console.log(`Array Pipeline:   ${t6.toFixed(4)}s`);
  console.log(`Set Ops:          ${t7.toFixed(4)}s`);
  console.log(`Promise Overhead: ${t8.toFixed(4)}s`);
  console.log("---------------");
  console.log(`TOTAL SCRIPT:     ${scriptTotal.toFixed(4)}s`);
  console.log();
  console.log(`RESULTS_CSV:${t1},${t2},${t3},${t4},${t5},${t6},${t7},${t8},${scriptTotal}`);
})();
